### PR TITLE
FIX-#2742: fix performance degradation for dictionary GroupBy aggregation

### LINKED ISSUE #2742: Groupby dictionary aggregation regression
### System information
- **OS Platform and Distribution (e.g., Linux Ubuntu 16.04)**: Ubuntu 20.04
- **Modin version** : aa818f5a246703621d4a70edaa04b92a511faa46
- **Python version**: 3.8.6
- **Code we can use to reproduce**:
```
import os

os.environ["MODIN_ENGINE"] = "ray"
os.environ["MODIN_CPUS"] = "4"

import modin.pandas as pd
from timeit import default_timer as timer

data_filename = "h2o_10k.csv"

df = pd.read_csv(data_filename)


def q1(df):
    groupby_cols = ["id1"]
    agg_cols_funcs = {"v1": "sum"}

    return df.groupby(groupby_cols).agg(agg_cols_funcs)

def q10(df):
    groupby_cols = ["id1", "id2", "id3", "id4", "id5", "id6"]
    agg_cols_funcs = {"v3": "sum", "v1": "count"}

    return df.groupby(groupby_cols).agg(agg_cols_funcs)

def meas_func(func, iters, *args, **kwargs):
    t_meas = float("inf")
    for _ in range(iters):
        t0 = timer()
        ans = func(*args, **kwargs)
        ans.shape
        t = timer() - t0
        del ans
        t_meas = t if t < t_meas else t_meas
    
    return t_meas

t_q1 = meas_func(q1, iters=3, **{"df": df})
t_q10 = meas_func(q10, iters=3, **{"df": df})

# results comparison between before vs after ee39d1767efb363fe6c147fc701b41315eee3afa commit
print("t_q1:", t_q1) # 0.026 s vs 0.031s
print("t_q10:", t_q10) # 0.091 s vs 2.09 s
```

### Describe the problem

Regression occurred after ee39d1767efb363fe6c147fc701b41315eee3afa commit.
data file [h2o_10k.txt](https://github.com/modin-project/modin/files/5988496/h2o_10k.txt)

### PR DESCRIPTION
**Pull request by dchigarev**
Signed-off-by: Dmitry Chigarev <dmitry.chigarev@intel.com>

## What do these changes do?

- [x] commit message follows format outlined [here](https://modin.readthedocs.io/en/latest/contributing.html)
- [x] passes `flake8 modin`
- [x] passes `black --check modin`
- [x] signed commit with `git commit -s` 
- [x] Resolves #2742 
- [x] tests pass

You can see the problem explanation in #2742. This PR replaces all the callables in [`groupby_reduce_functions`](https://github.com/modin-project/modin/blob/8ebbad9bca36aba0ddfc58dac89b9a66b0bfe41a/modin/data_management/functions/groupby_function.py#L234) to its string names, so now dictionary aggregation applies string functions to the columns and therefore performs much faster with a-lot-of-groups groupby. This PR also brings an ability to pass only a function name to the `GroupByReduce.register` for simplicity.

```
MODIN_CPUS=4 MODIN_TEST_DATASET_SIZE=Big asv continuous src/master HEAD --launch-method=spawn -b TimeGroupByDictionaryAggregation --split

       before           after         ratio
       [aa818f5a]       [c177b8aa]
         <master>         <issue-2742>
-        705±20ms         126±20ms     0.18  benchmarks.TimeGroupByDictionaryAggregation.time_groupby_dict_agg((5000, 5000), <function <lambda> at 0x7f37d61c8d40>, 'reduction')
-	 197±10ms         104±10ms     0.53  benchmarks.TimeGroupByDictionaryAggregation.time_groupby_dict_agg((1000000, 10), 100, 'reduction')
-	 944±20ms         113±10ms     0.12  benchmarks.TimeGroupByDictionaryAggregation.time_groupby_dict_agg((1000000, 10), <function <lambda> at 0x7f37d61c8d40>, 'reduction')
SOME BENCHMARKS HAVE CHANGED SIGNIFICANTLY.
PERFORMANCE INCREASED.
```

### CODE REVIEW

**Comment by anmyachev** [comment]
@dchigarev please split ASV changes into separate PR.

**Comment by amyskov** [comment]
Failed to find functional tests for case with multicolumn groupby and dictionary aggregation (are they really present?). If not, it will be good to extend test suite with this case.

**Comment by dchigarev** [comment]
@amyskov [`test_multi_column_groupby_different_partitions`](https://github.com/modin-project/modin/blob/8ebbad9bca36aba0ddfc58dac89b9a66b0bfe41a/modin/pandas/test/test_groupby.py#L1534) tests multi-column GroupBy with basic and dictionary aggregation

**Comment by amyskov** [comment]
> @amyskov [`test_multi_column_groupby_different_partitions`](https://github.com/modin-project/modin/blob/8ebbad9bca36aba0ddfc58dac89b9a66b0bfe41a/modin/pandas/test/test_groupby.py#L1534) tests multi-column GroupBy with basic and dictionary aggregation

Great, test suite is good enough.

**Comment by dchigarev** [comment]
@anmyachev @amyskov @devin-petersohn ready for review

File: modin/data_management/functions/groupby_function.py [hunk 1]
```diff
@@ -101,8 +133,8 @@ def reduce(
         if not as_index:
             insert_levels = partition_idx == 0 and (drop or method == "size")
             result.reset_index(drop=not insert_levels, inplace=True)
-
-        return result
+        # Result could not always be a frame, so wrapping it into DataFrame
```
Comment (anmyachev) [inline modin/data_management/functions/groupby_function.py:136]: Should we add check for this?

File: modin/data_management/functions/groupby_function.py [hunk 2]
```diff
@@ -63,7 +94,8 @@ def map(
         result = apply_func(
             df.groupby(by=by_part, axis=axis, **groupby_args), **map_args
         )
-        return result
+        # Result could not always be a frame, so wrapping it into DataFrame
```
Comment (anmyachev) [inline modin/data_management/functions/groupby_function.py:97]: Same about check type.

File: modin/data_management/functions/groupby_function.py (hunk 1 above)
Comment (dchigarev) [reply modin/data_management/functions/groupby_function.py:136]: check will be redundant, `pandas.DataFrame(obj)` converts `obj` to frame if `obj` is Series, and has no effect if `obj` is a DataFrame.

Comment (anmyachev) [reply modin/data_management/functions/groupby_function.py:136]: ok

**Review by anmyachev — APPROVED** [review]

**Comment by anmyachev** [comment]
@amyskov the ball is on your side

**Pull request merged**